crawl-ref/source/catch2-tests/test_mon-util.cc

```diff
@@ -26,3 +30,51 @@ TEST_CASE("can get names for removed monster types", "[single-file]" ) {
 
     REQUIRE(name == "removed bumblebee");
 }
+
+TEST_CASE("mons_habitat_type returns correct habitats", "[single-file]") {
+
+    SECTION("Land monsters should be HT_LAND") {
+        const habitat_type habitat = mons_habitat_type(MONS_KOBOLD,MONS_KOBOLD);
+
+        REQUIRE(habitat == HT_LAND);
+    }
+
+    SECTION("Lava monsters should be HT_LAVA") {
+        const habitat_type habitat = mons_habitat_type(MONS_LAVA_SNAKE,MONS_LAVA_SNAKE);
+
+        REQUIRE(habitat == HT_LAVA);
+    }
+    SECTION("Water monsters should be HT_WATER") {
+        const habitat_type habitat = mons_habitat_type(MONS_KRAKEN,MONS_KRAKEN);
+
+        REQUIRE(habitat == HT_WATER);
+    }
+
+    SECTION("Amphibious monsters should be HT_AMPHIBIOUS") {
+        const habitat_type habitat = mons_habitat_type(MONS_FRILLED_LIZARD,MONS_FRILLED_LIZARD);
+        REQUIRE(habitat == HT_AMPHIBIOUS);
+    }
+
+    SECTION("Amphibious (lava) monsters should be HT_AMPHIBIOUS_LAVA") {
+        const habitat_type habitat = mons_habitat_type(MONS_SALAMANDER,MONS_SALAMANDER);
+        REQUIRE(habitat == HT_AMPHIBIOUS_LAVA);
+    }
+
+    SECTION("Zombies of amphibious monsters should still be HT_AMPHIBIOUS") {
+        const habitat_type habitat = mons_habitat_type(MONS_ZOMBIE,MONS_FRILLED_LIZARD);
+        REQUIRE(habitat == HT_AMPHIBIOUS);
+    }
+
+    SECTION("Giants should show up as HT_AMPHIBIOUS unless explicitly told not to") {
+        habitat_type habitat = mons_habitat_type(MONS_CYCLOPS,MONS_CYCLOPS);
+        REQUIRE(habitat == HT_AMPHIBIOUS);
+
+        habitat = mons_habitat_type(MONS_CYCLOPS,MONS_CYCLOPS,true);
+        REQUIRE(habitat != HT_AMPHIBIOUS);
+    }
```

Comment:
nit: I'd recommend splitting this into two separate sections. It's best to have one assertion per section wherever possible.

Also: amphibious giants should still show up as amphibious even if `real_amphibious` is true, so that would be another good test case.